### ISSUE TITLE
.NET 6 DateOnly and TimeOnly mapping support

### ISSUE THREAD

**Issue by arnederuwe** [DapperLib/Dapper#1715]
.NET 6 will introduce the DateOnly and TimeOnly structs, these are good mapping candidates for SQL Server's Date and Time types.

There is an open issue in the .NET SqlClient repo [here](https://github.com/dotnet/SqlClient/issues/1009)

Am I correct to assume that once SqlClient supports it, Dapper will implicitly support it as well? Or is there some work required in this repo as well in order for this to work?

**Comment by kevingates** [comment]
Any timeline when this may be fixed?

**Comment by mgravell** [comment]
@kevingates see also #1728; there is an open branch, but we need to first be sure how this is intended to interact with the various providers

**Comment by szalapski** [comment]
Very much would like to see this.  It is weird that SQL Server's DATE type should map to anything but DateOnly in C#.

**Comment by h181422** [comment]
For those who are looking for a workaround: make your own type handler.

Add the following to your configuration:

`SqlMapper.AddTypeHandler(new SqlTimeOnlyTypeHandler());`

```
public class SqlTimeOnlyTypeHandler : SqlMapper.TypeHandler<TimeOnly>
{
    public override void SetValue(IDbDataParameter parameter, TimeOnly time)
    {
        parameter.Value = time.ToString();
    }

    public override TimeOnly Parse(object value)
    {
        return TimeOnly.FromTimeSpan((TimeSpan)value);
    }
}
```

**Comment by danielearwicker** [comment]
And similarly for `DateOnly`:

```cs
public class DapperSqlDateOnlyTypeHandler : SqlMapper.TypeHandler<DateOnly>
{
    public override void SetValue(IDbDataParameter parameter, DateOnly date)
        => parameter.Value = date.ToDateTime(new TimeOnly(0, 0));
    
    public override DateOnly Parse(object value)
        => DateOnly.FromDateTime((DateTime)value);
}
```

**Comment by K0rhak** [comment]
I had problems with database conversion so I had to specify the type. Like this

`public override void SetValue(IDbDataParameter parameter, DateOnly date)
 {
      parameter.DbType = DbType.DateTime;
      parameter.Value = date.ToDateTime(new TimeOnly(0, 0));
 }`

**Comment by PedroC88** [comment]
How exactly is the TypeHandler used? I have the following line in the constructor of my DB class 

```csharp
SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
SqlMapper.AddTypeHandler(new TimeOnlyTypeHandler());
```

and the following TypeHandlers
```csharp
public class DateOnlyTypeHandler : SqlMapper.TypeHandler<DateOnly>
{
    public override DateOnly Parse(object value) => DateOnly.FromDateTime((DateTime)value);

    public override void SetValue(IDbDataParameter parameter, DateOnly value)
    {
        parameter.DbType = DbType.Date;
        parameter.Value = value;
    }
}

public class TimeOnlyTypeHandler : SqlMapper.TypeHandler<TimeOnly>
{
    public override TimeOnly Parse(object value) => TimeOnly.FromDateTime((DateTime)value);

    public override void SetValue(IDbDataParameter parameter, TimeOnly value)
    {
        parameter.DbType = DbType.Time;
        parameter.Value = value;
    }
}
```

But I'm still getting a an exception that the Date field in my object is `NULL` and the breakpoints on the convertion methods aren't being hit.

**Comment by h181422** [comment]
``` c#
SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
SqlMapper.AddTypeHandler(new TimeOnlyTypeHandler());
```
I've got them in my container config, where they are registered at startup.

**Comment by PedroC88** [comment]
That makes more sense indeed. And I also realized that z.dapper.pluss doesn't use the TypeHandlers for bulk inserts, which is why the breakpoints weren't hitting the methods in the first place.

**Comment by ScottRFrost** [comment]
Found this issue in a search and was able to get the type handlers working (The SqlMapper.AddTypeHandler lines go in Startup.cs \ ConfigureServices if you're working on an API by they way, that took me a few minutes to figure out).

However, to make them work in converting either a MS SQL DateTime OR a MS SQL Time to TimeOnly, I needed this modification (MS SQL Date to DateOnly seemed to work without change):


        public class DateOnlyTypeHandler : SqlMapper.TypeHandler<DateOnly> // Dapper handler for DateOnly
        {
            public override DateOnly Parse(object value) => DateOnly.FromDateTime((DateTime)value);

            public override void SetValue(IDbDataParameter parameter, DateOnly value)
            {
                parameter.DbType = DbType.Date;
                parameter.Value = value;
            }
        }

        public class TimeOnlyTypeHandler : SqlMapper.TypeHandler<TimeOnly> // Dapper handler for TimeOnly
        {
            public override TimeOnly Parse(object value)
            {
                if (value.GetType() == typeof(DateTime))
                {
                    return TimeOnly.FromDateTime((DateTime)value);
                }
                else if (value.GetType() == typeof(TimeSpan))
                {
                    return TimeOnly.FromTimeSpan((TimeSpan)value);
                }
                return default;
            }

            public override void SetValue(IDbDataParameter parameter, TimeOnly value)
            {
                parameter.DbType = DbType.Time;
                parameter.Value = value;
            }
        }


Hopefully these mappings can be added to default dapper one day

**Comment by yacine-karim** [comment]
Make sure to declare the DateOnly property is declared as Nullable `DateOnly?`  if the table contains a null values

**Comment by AdisonCavani** [comment]
`SetValue` function is called and it works, however `Parse` is never called. It's always returning NULL;

**Comment by nhustak** [comment]
Is there still no native support for DateOnly & TimeOnly?

**Comment by mgravell** [comment]
There are some complications on the read side that make it a much bigger change than you would think. I have some ideas,though.

**Comment by KeithHenry** [comment]
To make this work with `record` classes I've been using an alternative constructor, which may be useful to others working around this...

With a select like: 

```sql
select convert(date, Created) 'day', count(*) 'value' 
from...
group by convert(date, Created)
```

This crashes:
```c#
public record WithDateOnly(DateOnly Day, int Value);
```

Because there is no constructor accepting `System.DateTime, System.Int32`, even though SQL has output its native `date` type.

The workaround is to add an alternative constructor:

```c#
public record WithDateOnly(DateOnly Day, int Value) {
    public WithDateOnly(DateTime day, int value) : 
        this(DateOnly.FromDateTime(day), value) { }
}
```

**Comment by VictorioBerra** [comment]
Just hit this today. Would love to see this built-in. DateOnly has been in 2 major .NET versions now.

**Comment by mgravell** [comment]
Untested (not at PC), but you could try:

``` c#
SqlMapper.AddTypeMap(typeof(DateOnly), (DbType)-1, true);
SqlMapper.AddTypeMap(typeof(TimeOnly), (DbType)-1, true);
```

It is hard for us to configure this automatically because different providers need different configurations to work correctly here.

**Comment by mkorsukov** [comment]
After dependencies upgrade to latest versions, new .NET types seem to work fine with Dapper.

```
<ItemGroup>
  <PackageReference Include="Dapper" Version="2.1.37" />
  <PackageReference Include="Microsoft.Data.SqlClient" Version="5.2.0" />
</ItemGroup>
```

And custom type handlers `SqlMapper.TypeHandler<DateOnly>` are not used by Dapper any more.

**Comment by nunutu29** [comment]
In some cases (i.e. System.Data.SqlClient) `parameter.DbType = DbType.Date` is automatically converted to DbType.DateTime ([See here](https://github.com/dotnet/corefx/blob/aeda6ab039f8d6e6041379ab156d51410b679c13/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlParameter.cs#L277)).

```cs
private class DateOnlyTypeHandler : SqlMapper.TypeHandler<DateOnly>
{
    public override void SetValue(IDbDataParameter parameter, DateOnly value)
    {
        // Sets the parameter's data type to DbType.Date.
        parameter.DbType = DbType.Date;

        // Note:
        // - In System.Data.SqlClient, setting DbType.Date is automatically converted to DbType.DateTime.
        // - In Microsoft.Data.SqlClient, setting remains DbType.Date.

        if (parameter.DbType == DbType.DateTime)
        {
            // If the parameter's data type has been changed to DbType.DateTime,
            // converts the DateOnly value to DateTime, setting the time to zero.
            parameter.Value = value.ToDateTime(new TimeOnly(0, 0, 0, 0));
        }
        else
        {
            // Otherwise, directly assigns the DateOnly value to the parameter.
            parameter.Value = value;
        }
    }

    public override DateOnly Parse(object value)
    {
        // Checks if the value is already of type DateOnly
        if (value is DateOnly dtOnly)
        {
            return dtOnly;
        }
        
        // Checks if the value is of type DateTime and converts it to DateOnly
        if (value is DateTime dt)
        {
            return DateOnly.FromDateTime(dt);
        }

        // If the value is neither DateOnly nor DateTime, throws an exception with an error message
        throw new InvalidOperationException($"Errore durante la conversione DateOnly: {value}" );
    }
}
```

**Comment by lansman** [comment]
@nunutu29 hi, I've an got an error in my project

> System.Data.DataException: Error parsing column 0 (date=23.09.2024 00:00:00 - DateTime) ---> System.InvalidCastException: Invalid cast from 'System.DateTime' to 'System.DateOnly'.

I use dotnet core 8.0. Dapper 2.1.35 (the latest nuget version). 
Database field is **Date**. C# DTO field is **DateOnly**. I don't want to have any intermediate _DateTime_ cast, how to avoid it?

> // - In System.Data.SqlClient, setting DbType.Date is automatically converted to DbType.DateTime.
> // - In Microsoft.Data.SqlClient, setting remains DbType.Date.

I don't use any of these libraries explicitly. I believe they are used by Dapper implicitly. But how can I find out which one is used? How can I switch to another?

**Comment by mgravell** [comment]
We're still trying to unpick the mess of DateOnly/TimeOnly; different ADO.NET providers have different demands here (and the underlying connection comes from *your* code - that isn't something that Dapper gets to choose); we're making progress on this topic over in DapperAOT, and when we've stabilised the rules there (where it is easier to debug etc), the plan is to migrate back as much as we can back into Dapper

**Comment by mkorsukov** [comment]
@lansman If you search through you projects (*.csproj files for "SqlClient"), you may find the actual SQL client library. Here is my case:

```xml
<ItemGroup>
  <PackageReference Include="Dapper" Version="2.1.44" />
  <PackageReference Include="Microsoft.Data.SqlClient" Version="5.2.2" />
</ItemGroup>
```

**Comment by lansman** [comment]
@mkorsukov I don't have any SqlClient in my project at all
![image](https://github.com/user-attachments/assets/2aadacb7-c89a-4ae7-9c1d-eb763e2a028f)

**Comment by buzz100** [comment]
I assume dapper will continue to support "System.Data.SqlClient", I personally don't need it but its now officially deprecated, this may be one less complication in the future.
 
https://techcommunity.microsoft.com/t5/sql-server-blog/announcement-system-data-sqlclient-package-is-now-deprecated/ba-p/4227205

**Comment by nunutu29** [comment]
> @nunutu29 hi, I've got an error in my project
> 
> > System.Data.DataException: Error parsing column 0 (date=23.09.2024 00:00:00 - DateTime) ---> System.InvalidCastException: Invalid cast from 'System.DateTime' to 'System.DateOnly'.
> 
> I use dotnet core 8.0. Dapper 2.1.35 (the latest nuget version). Npgsql 7.0.7. Npgsql.EntityFrameworkCore.PostgreSQL 7.0.18. UPD: upgraded to Npgsql, Npgsql.EntityFrameworkCore.PostgreSQL 8.0.4, Microsoft.EntityFrameworkCore 8.0.8, it didn't help. Database field is **Date**. C# DTO field is **DateOnly**. I don't want to have any intermediate _DateTime_ cast, how to avoid it?
> 
> > // - In System.Data.SqlClient, setting DbType.Date is automatically converted to DbType.DateTime.
> > // - In Microsoft.Data.SqlClient, setting remains DbType.Date.
> 
> I don't use any of these libraries explicitly. Are they used by Dapper implicitly? But how can I find out which one is used? How can I switch to another?

Hello @lansman,
The easiest way is to set a breakpoint during runtime and inspect the connection variable directly.

Additionally, to use `DateOnlyTypeHandler` make sure to include the following code in your startup class:

```cs
SqlMapper.RemoveTypeMap(typeof(DateOnly));
SqlMapper.AddTypeHandler(new DateOnlyTypeHandler());
```

**Comment by lansman** [comment]
@nunutu29
>The easiest way is to set a breakpoint during runtime and inspect the connection variable directly.
Still no clue
![image](https://github.com/user-attachments/assets/274f8289-52e9-456f-a92b-297020d02cf4)

> Additionally, to use DateOnlyTypeHandler make sure to include the following code in your startup class:
Thank you, it works for me!

**Comment by nunutu29** [comment]
> @nunutu29
> 
> > The easiest way is to set a breakpoint during runtime and inspect the connection variable directly.
> > Still no clue
> > ![image](https://private-user-images.githubusercontent.com/1588720/369962240-6760da58-0c86-44b6-9b9b-8c4339d1e1ae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxMTI3MzksIm5iZiI6MTcyNzExMjQzOSwicGF0aCI6Ii8xNTg4NzIwLzM2OTk2MjI0MC02NzYwZGE1OC0wYzg2LTQ0YjYtOWI5Yi04YzQzMzlkMWUxYWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjNUMTcyNzE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2QzNGFkYzU5YTUyZGE1NmE5MzY2OTk3OWM0OWNhODI4M2M0NTUxMDk1YjNlZDJlODY5YmU4MGQ4OTBmMmNmMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e0PPsEZ_oeV5j9rarU7CcQG6WGXxsvsK_u2odwy8Kpc)
> 
> > Additionally, to use DateOnlyTypeHandler make sure to include the following code in your startup class:
> 
> Thank you, it works for me!

The reason you don't find SqlClient it's simply because you don't have it. You are using [Npgsql](https://github.com/npgsql/npgsql).

**Comment by thuijzer** [comment]
For those looking for a simple solutions: in `net8.0` using `Dapper 2.1.35` and `MySql.Data 9.1.0` this is working for me (added in `Program.cs`):

```
SqlMapper.AddTypeMap(typeof(DateOnly), DbType.Date, true);
SqlMapper.AddTypeMap(typeof(TimeOnly), DbType.Time, true);
```

**Comment by mdsharpe** [comment]
Just had this error this morning after updating to version 2.1.66.

Reverting back to 2.1.44 solved it.

**Comment by sanjeevnofficial4324-collab** [comment]
During development, we identified an issue where DateOnly and DateOnly? fields were always being saved as NULL in PostgreSQL, even though the values were correctly populated in the application. Notably, Dapper did not throw any exception— it simply failed to map the DateOnly value to the database parameter.

This happened because Dapper does not natively support the .NET DateOnly type, and as a result, it could not translate the value into a format PostgreSQL understands (date).

To fix this, we implemented a custom Dapper type handler (SqlDateOnlyNullableTypeHandler), which explicitly converts:

DateOnly → DateTime when sending data to the database

DateTime → DateOnly when reading data from the database

After registering this type handler during application startup, Dapper began correctly mapping the DateOnly values, and the dates are now properly persisted in the database.

This solution is efficient, stable, and leverages Dapper’s extensibility model, ensuring accurate date handling across all insert/update operations without modifying the existing database schema.
